### ISSUE TITLE
win32 x86 build of openssl / visual studio 2012

### ISSUE THREAD

**Issue by glengillman** [openssl/openssl#3677]
Hi All,

I've been reading a lot of documentation and trying to step my way through building OpenSSL.

I've gone through the steps, installed Active Perl, installed NASM, then read a statement that it wasn't recommended to use that.  I am trying to build as x86.  So I opened Visual Studio 2012 command prompt - then I did the following:

NOTE:  I'm running this on a Windows 7 machine

ran this from the folder where the OpenSSL files were downloaded    (the prefix is an empty folder where I expected the target files to be created):

```
c:\perl\bin\perl.exe Configure VC-WIN32 --prefix=D:\Code_TFS\Shared\Libraries\FreeRDP\Main\RawFiles\OpenSSL_Solution
```

I observed that a new file was created named "makefile"

instructions say to run "do_ms" in the "ms" folder - but there are no executables in that folder, so I couldn't perform this step

```
nmake -f makefile
```

this ran for a bit.  There were several warnings and then errors.

Here are the first set of errors:     It seems that they are trying to utilize "Windows Kits\8.0" - and it is complaining about errors using that.   Is this the correct version of windows kits that should be used?   Is there any information about this ?   or how to install and/or configure it to use a different version ?

```
        cl  /I "." /I "include" -DOPENSSL_USE_APPLINK -DDSO_WIN32 -DNDEBUG -DOPENSSL_THREADS -DOPENSSL_NO_STATIC_ENGINE -DOPENSSL_PIC -DOPENSSL_BN_ASM_PART_WORDS -DOPENSSL_IA32_SSE2 -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_GF2m -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DRC4_ASM -DMD5_ASM -DRMD160_ASM -DAES_ASM -DVPAES_ASM -DWHIRLPOOL_ASM -DGHASH_ASM -DECP_NISTZ256_ASM -DPADLOCK_ASM -DPOLY1305_ASM "-DENGINESDIR=\"D:\\Code_TFS\\Shared\\Libraries\\FreeRDP\\Main\\RawFiles\\OpenSSL_Solution\\lib\\engines-1_1\"" "-DOPENSSLDIR=\"C:\\Program Files (x86)\\Common Files\\SSL\"" -W3 -wd4090 -Gs0 -GF -Gy -nologo -DOPENSSL_SYS_WIN32 -DWIN32_LEAN_AND_MEAN -DL_ENDIAN -D_CRT_SECURE_NO_DEPRECATE -DUNICODE -D_UNICODE /MD /O2 /Zi /Fdapp -c /Foapps\apps.obj "apps\apps.c"
apps.c
C:\Program Files (x86)\Windows Kits\8.0\include\um\windows.h(32) : warning C4067: unexpected tokens following preprocessor directive - expected a newline
apps\apps.c(2027) : error C2065: 'DWORD' : undeclared identifier
apps\apps.c(2027) : error C2146: syntax error : missing ';' before identifier 'err'
apps\apps.c(2027) : error C2065: 'err' : undeclared identifier
apps\apps.c(2028) : error C2143: syntax error : missing ';' before 'type'
apps\apps.c(2041) : warning C4013: 'MultiByteToWideChar' undefined; assuming extern returning int
apps\apps.c(2041) : error C2065: 'CP_ACP' : undeclared identifier
apps\apps.c(2041) : error C2065: 'WCHAR' : undeclared identifier
apps\apps.c(2041) : error C2059: syntax error : ')'
apps\apps.c(2043) : warning C4552: '<' : operator has no effect; expected operator with side-effect
apps\apps.c(2043) : error C2143: syntax error : missing ';' before ')'
apps\apps.c(2046) : error C2065: 'CP_ACP' : undeclared identifier
apps\apps.c(2046) : error C2065: 'WCHAR' : undeclared identifier
apps\apps.c(2046) : error C2059: syntax error : ')'
```

Any help is much appreciated.

Thank you

Glen

**Comment by levitte** [comment]
I have no idea what "Windows Kits" is, but it seems that you have added its include directory (or rather, a subdirectory of it) into Visual Studio's header file search path.  That needs correcting, as there's obviously a `windows.h` there that's not compatible with the standard Windows `windows.h`.

Seeing that the path it complains about is `C:\Program Files (x86)\Windows Kits\8.0\include\um\windows.h`, I would think that the path `C:\Program Files (x86)\Windows Kits\8.0\include\um` was added to that search path, and it really should just be `C:\Program Files (x86)\Windows Kits\8.0\include` (programs using that particular `windows.h` should probably include it like this: `#include <um/windows.h>`).

I cannot claim to know the workings of Visual Studio and its settings, but perhaps [this stackoverflow page](https://stackoverflow.com/questions/2676417/how-do-include-paths-work-in-visual-studio) can help you figure it out.

**Comment by levitte** [comment]
Side note: in the future, please tell us which OpenSSL version you're working with.  OpenSSL 1.1.*x* builds differently from 1.0.2.  Things like `ms\do_ms` is particular to pre 1.1.0 versions, while 1.1.0 simply produces the `makefile` you noticed.

**Comment by dot-asm** [comment]
`um` subdirectory in %INCLUDE% path is perfectly legitimate and it does work [for me]. For reference there are no header files in Windows Kits\*\include, just handful subdirectories, um being one of them. No, you shouldn't attempt to `#include <um/windows.h>`. Yes, Microsoft SDKs\Windows\*\include have header files, but that's different story. Yes, user can [figure out how to] switch to SDKs [by manipulating %INCLUDE% and %LIBPATH% environment variables], but Windows Kits should work, and does for me.

**Comment by dot-asm** [comment]
%INCLUDE% and %LIB%, not %LIBPATH%.

**Comment by dot-asm** [comment]
Key question is what's on line 32 of windows.h in question and why compiler gets angry about it. Not that I know answer to that question, nor that I offer myself to find one. After all we stand for code and user stands for *working* developer environment. We do make it work in default environment [like appveyor, but we can eventually confirm other versions], but we don't [have capacity to] get involved in every custom setting that cause some specific trouble.

This shouldn't discourage anybody to share their experience with similar problem. If it turns to be common problem, it might be worth considering it for FAQ...

**Comment by levitte** [comment]
Ok, shows how much I know...  sorry for my yacking

**Comment by mattcaswell** [comment]
You don't say what OpenSSL version, but I'm assuming 1.0.2.

> I've gone through the steps, installed Active Perl, installed NASM, then read a statement that it wasn't recommended to use that.

NASM is *the* recommended way to do this, so I'm not sure where you read that (was that in official docs - if so can you point at where it says that). Perhaps you are confusing it with MASM, which is not recommended.

> instructions say to run "do_ms" in the "ms" folder - but there are no executables in that folder, so I couldn't perform this step

The instructions say:

     - If you are using NASM then run:
       > ms\do_nasm

     - If you don't want to use the assembly language files at all then run:
       > perl Configure VC-WIN32 no-asm --prefix=c:/some/openssl/dir
       > ms\do_ms

So (since you didn't specify no-asm) you should be running `ms\do_nasm` not `ms\do_ms`. You must perform this step - if you can't perform it then you need to solve that problem before continuing. You say there are no executables in that folder. There should be several .bat files, including "do_ms.bat", "do_nasm.bat" etc. It is those that get executed by the above command. If they are not present then you have a corrupt distribution. Download it again to a fresh directory and start again.

I'm not sure if that is the cause of the error you posted above (it doesn't seem directly related), but certainly you need to solve these problems first.

**Comment by glengillman** [comment]
Thank you all for your prompt assistance.   FYI:  I was using OpenSSL version 1.1 because I wanted to use the latest version.    Regarding windows kits - I believe those are the C distributable version that was installed when I installed Visual Studio.  I have not modified anything there.  I wasn't sure if there was a specific version of these the Microsoft C libraries that should be used.  

NOTE:  I could have mistaken MASM for NASM - so you may disregard that comment.  If I find a reference, I'll definitely let you know.  

I'll definitely repeat the steps again with ms\do_nasm and maybe that will help.   You mentioned that if I don't want to use the assembly language files at all, then I can use the no-asm parameter.   What would you recommend - to use them or not use them?   It is not clear to me the value either way.

Thank you very much everyone

**Comment by mattcaswell** [comment]
> I was using OpenSSL version 1.1 

Ah, that changes things significantly. The build instructions are different between 1.0.2 and 1.1.0. It sounds like you are trying to apply 1.0.2 build procedure to 1.1.0. See the INSTALL file for details

Essentially the build procedure should be:

    > perl Configure VC-WIN32 --prefix=D:\Code_TFS\Shared\Libraries\FreeRDP\Main\RawFiles\OpenSSL_Solution
    > nmake
    > nmake test
    > nmake install

You don't need to run anything in the ms folder for a 1.1.0 build.

**Comment by glengillman** [comment]
Thank you everyone.  

What I learned is that the requirement to build this for 1.1:

Visual Studio 2010 Command Prompt
Windows SDK - installed C:\Program Files (x86)\Windows Kits\8.0

I was using Visual Studio 2012.  I looked on Microsoft websites, and newer versions of Visual studiodo not support this.   It is required to use Visual Studio 2010

It is building and the tests are passing.   Thank you so much.   Have a great day!
